In [139]:
# from datasets import load_market_data
import pandas as pd
# from modules._file_paths import OPTION_DATA
from modules._utils import market_holiday , holiday_url_2020
from modules import HistoricalData
import numpy as np
from datetime import datetime
from dateutil.relativedelta import relativedelta, TH
# from vasco.cc_financial_plots import general_plots
# from vasco import plot
import glob
import os
from tqdm import tqdm

In [4]:
eis_data = HistoricalData(source='eis_data', name = 'EIS DATA', underlying_instrument='NIFTY',start_date='20200101', end_date='20200101', expiry_type='nearest_weekly')

In [5]:
eis_data.load_market_data()

In [6]:
data = eis_data.getData()

In [11]:
data

ExchToken  BidPrice  BidQty  AskPrice  AskQty  TTq  LTP  \
Date Time                                                                      
2020-01-01 09:16:00      45994   2101.15    75.0   2119.55    75.0  0.0  0.0   
2020-01-01 09:17:00      45994   2092.50    75.0   2111.30    75.0  0.0  0.0   
2020-01-01 09:18:00      45994   2090.45    75.0   2109.25    75.0  0.0  0.0   
2020-01-01 09:19:00      45994   2093.85    75.0   2113.05    75.0  0.0  0.0   
2020-01-01 09:20:00      45994   2097.75    75.0   2116.80    75.0  0.0  0.0   
...                        ...       ...     ...       ...     ...  ...  ...   
2020-01-01 15:26:00      41620   1424.80   900.0   1651.90   900.0  0.0  0.0   
2020-01-01 15:27:00      41620   1424.80   900.0   1651.90   900.0  0.0  0.0   
2020-01-01 15:28:00      41620   1424.80   900.0   1651.90   900.0  0.0  0.0   
2020-01-01 15:29:00      41620   1424.80   900.0   1651.90   900.0  0.0  0.0   
2020-01-01 15:30:00      41620   1424.80   900.0   1651.90   900.0  0.0  0.0   

                     TotalTradedPrice Instrument ExpiryDate ExpiryTime  \
Date Time                                                                
2020-01-01 09:16:00               0.0      NIFTY 2020-01-02   15:30:00   
2020-01-01 09:17:00               0.0      NIFTY 2020-01-02   15:30:00   
2020-01-01 09:18:00               0.0      NIFTY 2020-01-02   15:30:00   
2020-01-01 09:19:00               0.0      NIFTY 2020-01-02   15:30:00   
2020-01-01 09:20:00               0.0      NIFTY 2020-01-02   15:30:00   
...                               ...        ...        ...        ...   
2020-01-01 15:26:00               0.0      NIFTY 2020-01-02   15:30:00   
2020-01-01 15:27:00               0.0      NIFTY 2020-01-02   15:30:00   
2020-01-01 15:28:00               0.0      NIFTY 2020-01-02   15:30:00   
2020-01-01 15:29:00               0.0      NIFTY 2020-01-02   15:30:00   
2020-01-01 15:30:00               0.0      NIFTY 2020-01-02   15:30:00   

                      Strike Type      ExpiryDateTime  
Date Time                                              
2020-01-01 09:16:00  10100.0   CE 2020-01-02 15:30:00  
2020-01-01 09:17:00  10100.0   CE 2020-01-02 15:30:00  
2020-01-01 09:18:00  10100.0   CE 2020-01-02 15:30:00  
2020-01-01 09:19:00  10100.0   CE 2020-01-02 15:30:00  
2020-01-01 09:20:00  10100.0   CE 2020-01-02 15:30:00  
...                      ...  ...                 ...  
2020-01-01 15:26:00  13750.0   PE 2020-01-02 15:30:00  
2020-01-01 15:27:00  13750.0   PE 2020-01-02 15:30:00  
2020-01-01 15:28:00  13750.0   PE 2020-01-02 15:30:00  
2020-01-01 15:29:00  13750.0   PE 2020-01-02 15:30:00  
2020-01-01 15:30:00  13750.0   PE 2020-01-02 15:30:00  

[41855 rows x 14 columns]

In [7]:
slice_data_obj = eis_data.getSlice('2020-01-01 09:16:00')

In [8]:
slice_data = slice_data_obj.getData()

In [9]:
slice_data[slice_data['Strike'] == 29100.0]

Empty DataFrame
Columns: [ExchToken, BidPrice, BidQty, AskPrice, AskQty, TTq, LTP, TotalTradedPrice, Instrument, ExpiryDate, ExpiryTime, Strike, Type, ExpiryDateTime]
Index: []

In [12]:
eis_data.get_nearest_strike_premium(t = '2020-01-01 09:16:00', strike = 29090)

array([[1205.8,  900. , 1948.8,  900. ]])

In [13]:
data['ExpiryDate'].unique()

array(['2020-01-02T00:00:00.000000000'], dtype='datetime64[ns]')

In [14]:
expiries = list(set(data['ExpiryDateTime']))
expiries

[Timestamp('2020-01-02 15:30:00')]

In [15]:
# eis_data.get_spot('2020-01-01 09:16:00')

In [16]:
# eis_data.get_spot_v2('2020-01-01 09:16:00')

In [17]:
# data.at['2020-01-01 09:16:00', 'BidPrice']

In [18]:
# eis_data.get_option_detail_from_id(45716)

In [19]:
eis_data.get_quote('2020-01-01 09:16:00', 'CE', '2020-01-02 15:30:00', 27500.0)

In [20]:
# eis_data.get_spot('2020-01-01 09:16:00')

In [21]:
eis_data.getData()['Type'][0]

'CE'

In [22]:
# trade_time = eis_data.getData().index[0].strftime('%Y-%m-%d %H:%M:%S')

In [23]:
# trade_times = [trade_time for trade_time in eis_data.getData().index.strftime('%Y-%m-%d %H:%M:%S')]

In [24]:
# for trade_time in trade_timeseis_data.get_exercise_list(qtime=trade_time)

In [25]:
# trade_time = eis_data.getData().index[0].strftime('%Y-%m-%d %H:%M:%S')
# expiries = eis_data.get_exercise_list(qtime=trade_time)

In [26]:
# expiries = eis_data.get_exercise_list(qtime=trade_time)
expiries = list(set(data['ExpiryDateTime']))

In [27]:
expiries

[Timestamp('2020-01-02 15:30:00')]

In [28]:
monthly_expiries = eis_data.get_specific_expiry(expiry_list = expiries, expiry_type = 'nearest_monthly')

In [29]:
monthly_expiries

In [30]:
weekly_expiries = eis_data.get_specific_expiry(expiry_list = expiries, expiry_type = 'weekly')

In [31]:
weekly_expiries

['2020-01-02']

In [32]:
eis_data.get_specific_expiry(expiry_list = expiries, expiry_type = 'nearest_weekly')

'2020-01-02'

In [33]:
eis_data.get_specific_expiry(expiry_list = expiries, expiry_type = 'nearest_monthly')

In [34]:
eis_data.get_specific_expiry(expiry_list = expiries, expiry_type = 'second_weekly')

In [35]:
eis_data.get_specific_expiry(expiry_list = expiries, expiry_type = 'second_monthly')

In [36]:
eis_data.get_specific_expiry(expiry_list = expiries, expiry_type = 'abc')

### Visualization

In [37]:
expiry_data = pd.DataFrame(columns=['Monthly_Expiry', 'Weekly_Expiry'])

In [40]:
expiry_data['Monthly_Expiry'] = [len(monthly_expiries)]
expiry_data['Weekly_Expiry'] = [len(weekly_expiries)]
expiry_type_count = expiry_data.T
expiry_type_count.rename(columns={expiry_type_count.columns[0]:'Frequency'}, inplace=True)
expiry_type_count.index.name = 'expiry_type'
expiry_type_count.reset_index(inplace=True)

In [41]:
import plotly.express as px

fig = px.bar(expiry_type_count, x="expiry_type", y=["Frequency"], title="Different Expiry Types")
fig.show()

In [ ]:
from vasco.cc_graph import count_plot
count_plot(data=data, feature='Type')

In [ ]:
# range of strike
count_plot(data=data, feature='Strike')

In [ ]:
# range of expiry
count_plot(data=data, feature='ExpiryDate')

### Make Specific FileFormat

In [157]:
# # old_file_names = glob.glob('/home/cloudcraftz/Github/Sample_NIFTY_EIS_DATA/2021/*.txt')
old_file_names = glob.glob('/home/cloudcraftz/Github/Sample_NIFTY_EIS_DATA/2021/*.txt')

# This function will rename the file as required (eg, From 'BANKNIFTY_20200101_HistoricData.csv' to 'BANKNIFTY_20200101_Intraday.csv')
def rename_files(old_file_names):
    '''
    Parameters :
    ----------
               old_file_names: file names with specific directory
    '''
    for old_file in old_file_names:
        new_file = old_file.replace('txt', 'csv')
        new_file = new_file.replace('HistoricData', 'Intraday')
        os.rename(old_file, new_file)


In [158]:
rename_files(old_file_names)

### Preprocess

- Add Header

In [159]:
# file_names = glob.glob('sample_data/*.csv')
# save_file_path = 'Updated_sample_data/'

file_names = glob.glob('/home/cloudcraftz/Github/Sample_NIFTY_EIS_DATA/2021/*.csv')
save_file_path = '/home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/2021/'

In [160]:
def initial_preprocess(file_names, save_file_path):
    
    for file_name in file_names:

        data = pd.read_csv(file_name)
        columns = ['Date Time','UnixTimefrom 1-1-1980','ExchToken','BidPrice','BidQty','AskPrice','AskQty',
                   'TTq','LTP','TotalTradedPrice','Instrument','ExpiryDate','ExpiryTime','Strike','Type']
        data = data.shift(1)
        first_row_values = data.columns.to_list()
        # need to handle case : when column values are same like 0.0 again 0.0 it will make it as 0.0.1
        corrected_first_row_values = []
        for v in first_row_values:
            if v.count('.') == 2:
                corrected_first_row_values.append(v[:-2])
            else:
                corrected_first_row_values.append(v)
        data.loc[0] = corrected_first_row_values
        data.columns = columns
        data.to_csv(save_file_path+file_name.split('/')[-1], index = False)

In [161]:
initial_preprocess(file_names, save_file_path)

In [163]:
pd.read_csv('/home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/2021/NIFTY_20210101_Intraday.csv')

Date Time  UnixTimefrom 1-1-1980  ExchToken  BidPrice  \
0       2021-01-01 09:16:00           1.293960e+18    37930.0   66885.0   
1       2021-01-01 09:17:00           1.293960e+18    37930.0  159945.0   
2       2021-01-01 09:18:00           1.293960e+18    37930.0  159945.0   
3       2021-01-01 09:19:00           1.293960e+18    37930.0   66885.0   
4       2021-01-01 09:20:00           1.293960e+18    37930.0   66885.0   
...                     ...                    ...        ...       ...   
582249  2021-01-01 15:25:00           1.293965e+18    49156.0  120110.0   
582250  2021-01-01 15:26:00           1.293965e+18    49156.0  120110.0   
582251  2021-01-01 15:27:00           1.293965e+18    49156.0  120110.0   
582252  2021-01-01 15:28:00           1.293965e+18    49156.0  120110.0   
582253  2021-01-01 15:29:00           1.293965e+18    49156.0  120110.0   

        BidQty  AskPrice  AskQty  TTq  LTP  TotalTradedPrice Instrument  \
0        500.0  279110.0  2400.0  0.0  0.0               0.0      NIFTY   
1        750.0  279110.0  2400.0  0.0  0.0               0.0      NIFTY   
2        750.0  279110.0  2400.0  0.0  0.0               0.0      NIFTY   
3        500.0  279110.0  2400.0  0.0  0.0               0.0      NIFTY   
4        500.0  279110.0  2400.0  0.0  0.0               0.0      NIFTY   
...        ...       ...     ...  ...  ...               ...        ...   
582249  4875.0  180775.0  4875.0  0.0  0.0               0.0      NIFTY   
582250  4875.0  180775.0  4875.0  0.0  0.0               0.0      NIFTY   
582251  4875.0  180775.0  4875.0  0.0  0.0               0.0      NIFTY   
582252  4875.0  180775.0  4875.0  0.0  0.0               0.0      NIFTY   
582253  4875.0  180775.0  4875.0  0.0  0.0               0.0      NIFTY   

         ExpiryDate ExpiryTime     Strike Type  
0        30-12-2021   15:30:00  1380000.0   PE  
1        30-12-2021   15:30:00  1380000.0   PE  
2        30-12-2021   15:30:00  1380000.0   PE  
3        30-12-2021   15:30:00  1380000.0   PE  
4        30-12-2021   15:30:00  1380000.0   PE  
...             ...        ...        ...  ...  
582249   25-03-2021   15:30:00  1545000.0   PE  
582250   25-03-2021   15:30:00  1545000.0   PE  
582251   25-03-2021   15:30:00  1545000.0   PE  
582252   25-03-2021   15:30:00  1545000.0   PE  
582253   25-03-2021   15:30:00  1545000.0   PE  

[582254 rows x 15 columns]

### Date Difference

In [137]:
from datetime import datetime
from modules._utils import date_difference

In [27]:
cur_date = data.index[0]
cur_date

Timestamp('2021-03-10 09:16:00')

In [23]:
monthly_exp =pd.to_datetime(monthly_expiries)
monthly_exp

Timestamp('2020-01-30 00:00:00')

In [24]:
date_difference(cur_date, monthly_exp)

28

### All File Strike and Expiry Date check

In [169]:
save_file_path = '/home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/2021/*.csv'

In [170]:
files = glob.glob(save_file_path)

In [171]:
files

['/home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/2021/NIFTY_20210112_Intraday.csv',
 '/home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/2021/NIFTY_20210618_Intraday.csv',
 '/home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/2021/NIFTY_20210416_Intraday.csv',
 '/home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/2021/NIFTY_20210203_Intraday.csv',
 '/home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/2021/NIFTY_20210806_Intraday.csv',
 '/home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/2021/NIFTY_20210304_Intraday.csv',
 '/home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/2021/NIFTY_20210315_Intraday.csv',
 '/home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/2021/NIFTY_20210823_Intraday.csv',
 '/home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/2021/NIFTY_20211227_Intraday.csv',
 '/home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/2021/NIFTY_20211022_Intraday.csv',
 '/home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/2021/NIFTY_20210915_Intraday.csv',
 '/home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/2021/NIFTY_2021

In [172]:
len(files)

235

In [173]:
store_final_df = pd.DataFrame(columns = ['Date', 'filename', 'Strike', 'ExpiryDate', 'number_of_strike', 'number_of_expiry'])
not_done = []
for file in tqdm(files):
    try:
        
        file_name = file.split('/')[-1]
        date = file_name.split('_')[1]
        date_format = date[:4] + '-' + date[4:6]+'-' +date[6:]
        eis_data = HistoricalData(source='eis_data', name = 'EIS DATA', underlying_instrument='NIFTY',start_date=date, end_date=date, expiry_type='all')

        eis_data.load_market_data()
        data = eis_data.getData()

        expiries = data['ExpiryDate'].unique()
        #expiries = pd.to_datetime(expiries).strftime('%Y-%m-%d')
        expiries = [expiry for expiry in pd.to_datetime(data['ExpiryDate'].unique()).strftime('%Y-%m-%d')]

        store_df = pd.DataFrame(columns=['Date', 'filename', 'Strike', 'ExpiryDate', 'number_of_strike', 'number_of_expiry'])
        store_df['Date'] = [date_format]
        store_df ['filename'] = [file_name]
        store_df['Strike'] = [data['Strike'].unique().tolist()]
        store_df['ExpiryDate'] = [expiries]
        store_df['number_of_strike'] = [len(data['Strike'].unique().tolist())]
        store_df['number_of_expiry'] = [len(expiries)]

    #     #assign to final data
        store_final_df = store_final_df.append(store_df)
    except Exception as e:
        print(e)
        not_done.append(file.split('/')[-1])

 64%|█████████████████████████▌              | 150/235 [00:00<00:00, 746.20it/s]

Data file not found: /home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/NIFTY_20210112_Intraday.csv
Data file not found: /home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/NIFTY_20210618_Intraday.csv
Data file not found: /home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/NIFTY_20210416_Intraday.csv
Data file not found: /home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/NIFTY_20210203_Intraday.csv
Data file not found: /home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/NIFTY_20210806_Intraday.csv
Data file not found: /home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/NIFTY_20210304_Intraday.csv
Data file not found: /home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/NIFTY_20210315_Intraday.csv
Data file not found: /home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/NIFTY_20210823_Intraday.csv
Data file not found: /home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/NIFTY_20211227_Intraday.csv
Data file not found: /home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/NIFTY_20211022_Intraday.csv
Data file not found: /home/clo

100%|████████████████████████████████████████| 235/235 [00:00<00:00, 756.09it/s]

Data file not found: /home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/NIFTY_20210504_Intraday.csv
Data file not found: /home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/NIFTY_20210208_Intraday.csv
Data file not found: /home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/NIFTY_20210503_Intraday.csv
Data file not found: /home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/NIFTY_20210907_Intraday.csv
Data file not found: /home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/NIFTY_20210310_Intraday.csv
Data file not found: /home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/NIFTY_20210818_Intraday.csv
Data file not found: /home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/NIFTY_20210510_Intraday.csv
Data file not found: /home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/NIFTY_20211026_Intraday.csv
Data file not found: /home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/NIFTY_20211116_Intraday.csv
Data file not found: /home/cloudcraftz/Github/NIFTY_EIS_DATA_Updated/NIFTY_20210420_Intraday.csv
Data file not found: /home/clo

In [104]:
store_final_df.head()

Date                     filename  \
0  2020-02-13  NIFTY_20200213_Intraday.csv   
0  2020-06-30  NIFTY_20200630_Intraday.csv   
0  2020-04-22  NIFTY_20200422_Intraday.csv   
0  2020-04-27  NIFTY_20200427_Intraday.csv   
0  2020-06-24  NIFTY_20200624_Intraday.csv   

                                              Strike  \
0  [12000.0, 12500.0, 11500.0, 13000.0, 11000.0, ...   
0  [10400.0, 10600.0, 12000.0, 12500.0, 10800.0, ...   
0  [12000.0, 13700.0, 5900.0, 12500.0, 12600.0, 1...   
0  [12000.0, 13700.0, 5900.0, 12500.0, 12600.0, 1...   
0  [10400.0, 10600.0, 12000.0, 13700.0, 5900.0, 1...   

                                          ExpiryDate number_of_strike  \
0  [2020-06-25, 2020-12-31, 2020-03-26, 2020-02-2...              104   
0  [2020-12-31, 2020-09-24, 2021-03-25, 2020-07-0...              144   
0  [2020-06-25, 2020-12-31, 2020-09-24, 2020-04-3...              179   
0  [2020-06-25, 2020-12-31, 2020-09-24, 2020-04-3...              179   
0  [2020-12-31, 2020-06-25, 2020-09-24, 2021-03-2...              157   

  number_of_expiry  
0               12  
0               12  
0               12  
0               12  
0               13

In [105]:
save_df = store_final_df.set_index('Date')

In [106]:
save_df.index = pd.to_datetime(save_df.index)
save_df.sort_index(inplace=True)

In [83]:
# save_df.loc['2020-08-03']

In [84]:
# save_df.to_csv('/home/cloudcraftz/Github/Store_all_expiry_strike/eis_data_2020.csv')

In [85]:
# save_df = pd.read_csv('/home/cloudcraftz/Github/Store_all_expiry_strike/eis_data_2020.csv', index_col=0)
# save_df.index = pd.to_datetime(save_df.index)

In [107]:
from vasco.cc_graph import count_plot
count_plot(data=save_df, feature='number_of_strike',)

alt.LayerChart(...)

In [108]:
from vasco.cc_graph import count_plot
count_plot(data=save_df, feature='number_of_expiry')

alt.LayerChart(...)

In [120]:
save_df['number_of_strike'] = save_df['number_of_strike'].astype(int)
save_df['number_of_expiry'] = save_df['number_of_expiry'].astype(int)

### Find out which days have 0 strikes.

In [126]:
zero_strike = save_df[save_df['number_of_strike'] == 0]
zero_strike

Empty DataFrame
Columns: [filename, Strike, ExpiryDate, number_of_strike, number_of_expiry]
Index: []

### Find out Expiry less than 5 

In [127]:
expiry_less_than_five = save_df[save_df['number_of_expiry'] < 5]
expiry_less_than_five

Empty DataFrame
Columns: [filename, Strike, ExpiryDate, number_of_strike, number_of_expiry]
Index: []

### Find out Correct File :

    If day is not expiry day, nearest weekly and nearest monthly 
    
    If day is an expiry day((thursday) then nearest_weekly, nearest_monthly, second_weekly, second_monthly)

In [23]:
from datetime import datetime
from modules._utils import date_difference

dates = expiry_less_than_five.index
c = 0

thursday_corrput_files_with_len_more_than_three = []
thursday_corrput_files_with_len_less_than_four = []

not_thursday_corrput_files_with_len_more_than_one = []
not_thursday_corrput_files_with_len_less_than_two = []

thrusday_correct_files = []
not_thrusday_correct_files = []

exception_files = []
# If today is Thursday (1 = Mon, 2 = Tue, 3 = Wen ...)
for idx, val in tqdm(enumerate(expiry_less_than_five['filename'].values)):
    
    try:
        
        file_name = val.split('/')[-1]
        date = file_name.split('_')[1]
        date_format = date[:4] + '-' + date[4:6]+'-' +date[6:]
        
        if expiry_less_than_five.index[idx].isoweekday() == 4:
            # if thursday, we should have nearest_weekly, nearest_monthly, second_weekly , second_monthly
            expiry_dates_list = [pd.to_datetime(expiry) for expiry in expiry_less_than_five['ExpiryDate'].loc[dates[c]]]
            
            if len(expiry_dates_list) > 3:
                # 
                nearest_weekly_eis_data = HistoricalData(source='eis_data', name = 'EIS DATA', underlying_instrument='BANKNIFTY',start_date=date, end_date=date, expiry_type='nearest_weekly')
                nearest_weekly_eis_data.load_market_data()
                nearest_weekly = eis_data.getData()

                nearest_monthly_eis_data = HistoricalData(source='eis_data', name = 'EIS DATA', underlying_instrument='BANKNIFTY',start_date=date, end_date=date, expiry_type='nearest_monthly')
                nearest_monthly_eis_data.load_market_data()
                nearest_monthly = eis_data.getData()

                second_weekly_eis_data = HistoricalData(source='eis_data', name = 'EIS DATA', underlying_instrument='BANKNIFTY',start_date=date, end_date=date, expiry_type='second_weekly')
                second_weekly_eis_data.load_market_data()
                second_weekly = eis_data.getData()

                second_monthly_eis_data = HistoricalData(source='eis_data', name = 'EIS DATA', underlying_instrument='BANKNIFTY',start_date=date, end_date=date, expiry_type='second_monthly')
                second_monthly_eis_data.load_market_data()
                second_monthly = eis_data.getData()

                if len(nearest_weekly) == 0 or len(nearest_monthly) == 0 or len(second_weekly) == 0 or len(second_monthly) == 0:
                    thursday_corrput_files_with_len_more_than_three.append(val)
                else:
                    thrusday_correct_files.append(val)

            else:
                #save corrupt file
                thursday_corrput_files_with_len_less_than_four.append(val)

            c += 1

        else:
            # if not thursday , we should have nearest weekly and nearest monthly
            expiry_dates_list = [pd.to_datetime(expiry) for expiry in expiry_less_than_five['ExpiryDate'].loc[dates[c]]]
            if len(expiry_dates_list) > 1:
                # need to check dates expiry dates are nearest_weekly, nearest_monthly

                nearest_weekly_eis_data = HistoricalData(source='eis_data', name = 'EIS DATA', underlying_instrument='BANKNIFTY',start_date=date, end_date=date, expiry_type='nearest_weekly')
                nearest_weekly_eis_data.load_market_data()
                nearest_weekly = eis_data.getData()

                nearest_monthly_eis_data = HistoricalData(source='eis_data', name = 'EIS DATA', underlying_instrument='BANKNIFTY',start_date=date, end_date=date, expiry_type='nearest_monthly')
                nearest_monthly_eis_data.load_market_data()
                nearest_monthly = eis_data.getData()

                if len(nearest_weekly) == 0 or len(nearest_monthly) == 0:

                    not_thursday_corrput_files_with_len_more_than_one.append(val)
                else:
                    not_thrusday_correct_files.append(val)

            else:
                # #save corrupt file
                not_thursday_corrput_files_with_len_less_than_two.append(val)
            c += 1
    except Exception as e:
        print(e)
        print(val)
        exception_files.append(val)

13it [00:53,  2.88s/it]

list index out of range
BANKNIFTY_20200731_Intraday.csv


14it [00:56,  2.78s/it]

list index out of range
BANKNIFTY_20200803_Intraday.csv


15it [00:58,  2.70s/it]

list index out of range
BANKNIFTY_20200804_Intraday.csv


16it [01:01,  2.62s/it]

list index out of range
BANKNIFTY_20200805_Intraday.csv


18it [01:02,  1.76s/it]

list index out of range
BANKNIFTY_20200807_Intraday.csv


19it [01:03,  1.63s/it]

list index out of range
BANKNIFTY_20200810_Intraday.csv


20it [01:04,  1.53s/it]

list index out of range
BANKNIFTY_20200811_Intraday.csv


21it [01:06,  1.46s/it]

list index out of range
BANKNIFTY_20200812_Intraday.csv
list index out of range
BANKNIFTY_20200814_Intraday.csv


24it [01:07,  1.13it/s]

list index out of range
BANKNIFTY_20200817_Intraday.csv


25it [01:08,  1.07it/s]

list index out of range
BANKNIFTY_20200818_Intraday.csv


26it [01:09,  1.03it/s]

list index out of range
BANKNIFTY_20200819_Intraday.csv


103it [01:14, 15.38it/s]

list index out of range
BANKNIFTY_20201216_Intraday.csv


105it [01:15, 10.55it/s]

list index out of range
BANKNIFTY_20201218_Intraday.csv


106it [01:17,  7.32it/s]

list index out of range
BANKNIFTY_20201221_Intraday.csv


107it [01:19,  5.24it/s]

list index out of range
BANKNIFTY_20201223_Intraday.csv


110it [01:26,  1.28it/s]


In [24]:
thursday_corrput_files_with_len_more_than_three

[]

Missing Info:
-----------------

Assumption : All nearest_weekly expiry files are valid for weekly expiry options.

Number of Strikes 0 Files : 
============
'BANKNIFTY_20200319_Intraday.csv', 'BANKNIFTY_20200814_Intraday.csv'

Expiry Day (Thrusday) Files :
===================

  (1) 'BANKNIFTY_20200730_Intraday.csv', (second_monthly does not exists)

  (2) 'BANKNIFTY_20200806_Intraday.csv' (nearest weekly , second weekly exists but nearest monthly does not exists)

  (3) 'BANKNIFTY_20200813_Intraday.csv' (only nearest weekly exists)

  (4) 'BANKNIFTY_20200820_Intraday.csv' (only nearest weekly exists)

  (5) 'BANKNIFTY_20200827_Intraday.csv' (second monthly does not exists) (number of expiry is 1 i.e, 2020-08-27)

  (6) 'BANKNIFTY_20200903_Intraday.csv' (only nearest weekly exists)

  (7) 'BANKNIFTY_20200910_Intraday.csv' (only nearest weekly exists)

  (8) 'BANKNIFTY_20200917_Intraday.csv' (only nearest weekly exists)

  (9) 'BANKNIFTY_20200924_Intraday.csv' (only nearest weekly exists)

  (10) 'BANKNIFTY_20201001_Intraday.csv' (only nearest weekly exists)

  (11) 'BANKNIFTY_20201008_Intraday.csv' (only nearest weekly exists)

  (12) 'BANKNIFTY_20201015_Intraday.csv' (only nearest weekly exists)

  (13) 'BANKNIFTY_20201022_Intraday.csv' (only nearest weekly exists)

  (14) 'BANKNIFTY_20201029_Intraday.csv' (only nearest weekly exists)

  (15) 'BANKNIFTY_20201105_Intraday.csv' (only nearest weekly exists)

  (16) 'BANKNIFTY_20201112_Intraday.csv' (only nearest weekly exists)

  (17) 'BANKNIFTY_20201119_Intraday.csv' (only nearest weekly exists)

  (18) 'BANKNIFTY_20201203_Intraday.csv' (only nearest weekly exists)

  (19) 'BANKNIFTY_20201210_Intraday.csv' (only nearest weekly exists)

  (20) 'BANKNIFTY_20201217_Intraday.csv' (only nearest weekly exists)


Other than Expiry Day (thursday) Files:
===========================

  (1) 'BANKNIFTY_20200825_Intraday.csv' (is correct file? only one expiry exists-'2020-08-27')

  (2) 'BANKNIFTY_20200826_Intraday.csv'  (is correct file ?  only one expiry exists - '2020-08-27')

  (3) ''BANKNIFTY_20200828_Intraday.csv', (only nearest weekly exists)

  (4) 'BANKNIFTY_20200831_Intraday.csv' (only nearest weeekly exists)

  (5) 'BANKNIFTY_20200901_Intraday.csv', (only nearest weekly exists)

  (6) 'BANKNIFTY_20200902_Intraday.csv' (only nearest weekly exists)

  (7) 'BANKNIFTY_20200904_Intraday.csv', (only nearest weekly exists)

  (8)'BANKNIFTY_20200907_Intraday.csv',  (only nearest weekly exists)

  (9)'BANKNIFTY_20200908_Intraday.csv',  (only nearest weekly exists)

  (10)'BANKNIFTY_20200909_Intraday.csv' (only nearest weekly exists)

  (11)'BANKNIFTY_20200911_Intraday.csv',  (only nearest weekly exists)

  (12)'BANKNIFTY_20200914_Intraday.csv',  (only nearest weekly exists)

  (13)'BANKNIFTY_20200915_Intraday.csv' (only nearest weekly exists)

  (14)'BANKNIFTY_20200916_Intraday.csv' (only nearest weekly exists)

  (15)'BANKNIFTY_20200918_Intraday.csv',  (only nearest weekly exists)

  (16)'BANKNIFTY_20200921_Intraday.csv',  (only nearest weekly exists)

  (17)'BANKNIFTY_20200922_Intraday.csv' (only nearest weekly exists)

  (18)'BANKNIFTY_20200923_Intraday.csv' (only nearest weekly exists)

  (19)'BANKNIFTY_20200925_Intraday.csv',  (only nearest weekly exists)

  (20)'BANKNIFTY_20200928_Intraday.csv',  (only nearest weekly exists)

  (21)'BANKNIFTY_20200929_Intraday.csv' (only nearest weekly exists)

  (22)'BANKNIFTY_20200930_Intraday.csv' (only nearest weekly exists)

  (23)'BANKNIFTY_20201005_Intraday.csv',  (only nearest weekly exists)

  (24)'BANKNIFTY_20201006_Intraday.csv' (only nearest weekly exists)

  (25)'BANKNIFTY_20201007_Intraday.csv' (only nearest weekly exists)

  (26)'BANKNIFTY_20201009_Intraday.csv' (only nearest weekly exists)

  (27)'BANKNIFTY_20201012_Intraday.csv',  (only nearest weekly exists)

  (28)'BANKNIFTY_20201013_Intraday.csv' (only nearest weekly exists)

  (29)'BANKNIFTY_20201014_Intraday.csv' (only nearest weekly exists)

  (30)'BANKNIFTY_20201016_Intraday.csv' (only nearest weekly exists)

  (31)'BANKNIFTY_20201019_Intraday.csv',  (only nearest weekly exists)

  (32)'BANKNIFTY_20201020_Intraday.csv' (only nearest weekly exists)

  (33)'BANKNIFTY_20201021_Intraday.csv' (only nearest weekly exists)

  (34)'BANKNIFTY_20201023_Intraday.csv',  (only nearest weekly exists)

  (35)'BANKNIFTY_20201027_Intraday.csv' (only nearest weekly exists)

  (36)'BANKNIFTY_20201030_Intraday.csv' (only nearest weekly exists)

  (37)'BANKNIFTY_20201102_Intraday.csv' (only nearest weekly exists)

  (38)'BANKNIFTY_20201103_Intraday.csv',  (only nearest weekly exists)

  (39)'BANKNIFTY_20201104_Intraday.csv' (only nearest weekly exists)

  (40)'BANKNIFTY_20201106_Intraday.csv' (only nearest weekly exists)
  
  (41)'BANKNIFTY_20201109_Intraday.csv',  (only nearest weekly exists)

  (42)'BANKNIFTY_20201110_Intraday.csv' (only nearest weekly exists)

  (43)'BANKNIFTY_20201111_Intraday.csv' (only nearest weekly exists) 
  
  (44)'BANKNIFTY_20201113_Intraday.csv' (only nearest weekly exists)

  (45)'BANKNIFTY_20201117_Intraday.csv',  (only nearest weekly exists)

  (46)'BANKNIFTY_20201118_Intraday.csv' (only nearest weekly exists)

  (47)'BANKNIFTY_20201111_Intraday.csv' (only nearest weekly exists) 
  
  (48)'BANKNIFTY_20201120_Intraday.csv' (Is it correct file ?only one expiry exists-  2020-11-26)

  (49)'BANKNIFTY_20201125_Intraday.csv' (Is it correct file ?only one expiry exists- 2020-11-26)      

  (50) 'BANKNIFTY_20201127_Intraday.csv' (only nearest weekly exists)

  (51)  'BANKNIFTY_20201201_Intraday.csv',(only nearest weekly exists) 

  (52) 'BANKNIFTY_20201202_Intraday.csv',(only nearest weekly exists) 

  (53) 'BANKNIFTY_20201204_Intraday.csv',(only nearest weekly exists)

  (54) 'BANKNIFTY_20201207_Intraday.csv',(only nearest weekly exists)

  (55) 'BANKNIFTY_20201208_Intraday.csv',(only nearest weekly exists)

  (56) 'BANKNIFTY_20201209_Intraday.csv'(only nearest weekly exists)

  (57) 'BANKNIFTY_20201211_Intraday.csv',(only nearest weekly exists)

  (58) 'BANKNIFTY_20201214_Intraday.csv',(only nearest weekly exists)

  (59) 'BANKNIFTY_20201215_Intraday.csv'(only nearest weekly exists)


In [25]:
thursday_corrput_files_with_len_less_than_four

['BANKNIFTY_20200319_Intraday.csv',
 'BANKNIFTY_20200730_Intraday.csv',
 'BANKNIFTY_20200806_Intraday.csv',
 'BANKNIFTY_20200813_Intraday.csv',
 'BANKNIFTY_20200820_Intraday.csv',
 'BANKNIFTY_20200827_Intraday.csv',
 'BANKNIFTY_20200903_Intraday.csv',
 'BANKNIFTY_20200910_Intraday.csv',
 'BANKNIFTY_20200917_Intraday.csv',
 'BANKNIFTY_20200924_Intraday.csv',
 'BANKNIFTY_20201001_Intraday.csv',
 'BANKNIFTY_20201008_Intraday.csv',
 'BANKNIFTY_20201015_Intraday.csv',
 'BANKNIFTY_20201022_Intraday.csv',
 'BANKNIFTY_20201029_Intraday.csv',
 'BANKNIFTY_20201105_Intraday.csv',
 'BANKNIFTY_20201112_Intraday.csv',
 'BANKNIFTY_20201119_Intraday.csv',
 'BANKNIFTY_20201203_Intraday.csv',
 'BANKNIFTY_20201210_Intraday.csv',
 'BANKNIFTY_20201217_Intraday.csv']

In [26]:
not_thursday_corrput_files_with_len_more_than_one

[]

In [27]:
# other than thrusday corrput files which has less than two expiries
not_thursday_corrput_files_with_len_less_than_two 

['BANKNIFTY_20200825_Intraday.csv',
 'BANKNIFTY_20200826_Intraday.csv',
 'BANKNIFTY_20200828_Intraday.csv',
 'BANKNIFTY_20200831_Intraday.csv',
 'BANKNIFTY_20200901_Intraday.csv',
 'BANKNIFTY_20200902_Intraday.csv',
 'BANKNIFTY_20200904_Intraday.csv',
 'BANKNIFTY_20200907_Intraday.csv',
 'BANKNIFTY_20200908_Intraday.csv',
 'BANKNIFTY_20200909_Intraday.csv',
 'BANKNIFTY_20200911_Intraday.csv',
 'BANKNIFTY_20200914_Intraday.csv',
 'BANKNIFTY_20200915_Intraday.csv',
 'BANKNIFTY_20200916_Intraday.csv',
 'BANKNIFTY_20200918_Intraday.csv',
 'BANKNIFTY_20200921_Intraday.csv',
 'BANKNIFTY_20200922_Intraday.csv',
 'BANKNIFTY_20200923_Intraday.csv',
 'BANKNIFTY_20200925_Intraday.csv',
 'BANKNIFTY_20200928_Intraday.csv',
 'BANKNIFTY_20200929_Intraday.csv',
 'BANKNIFTY_20200930_Intraday.csv',
 'BANKNIFTY_20201005_Intraday.csv',
 'BANKNIFTY_20201006_Intraday.csv',
 'BANKNIFTY_20201007_Intraday.csv',
 'BANKNIFTY_20201009_Intraday.csv',
 'BANKNIFTY_20201012_Intraday.csv',
 'BANKNIFTY_20201013_Intrada

In [28]:
thrusday_correct_files

['BANKNIFTY_20200723_Intraday.csv']

In [29]:
not_thrusday_correct_files

['BANKNIFTY_20200323_Intraday.csv',
 'BANKNIFTY_20200717_Intraday.csv',
 'BANKNIFTY_20200720_Intraday.csv',
 'BANKNIFTY_20200721_Intraday.csv',
 'BANKNIFTY_20200722_Intraday.csv',
 'BANKNIFTY_20200724_Intraday.csv',
 'BANKNIFTY_20200727_Intraday.csv',
 'BANKNIFTY_20200728_Intraday.csv',
 'BANKNIFTY_20200729_Intraday.csv',
 'BANKNIFTY_20200821_Intraday.csv',
 'BANKNIFTY_20200824_Intraday.csv',
 'BANKNIFTY_20201228_Intraday.csv',
 'BANKNIFTY_20201229_Intraday.csv',
 'BANKNIFTY_20201230_Intraday.csv']

In [30]:
exception_files # nearest monthly doesnot exists or only expiry date available

['BANKNIFTY_20200731_Intraday.csv',
 'BANKNIFTY_20200803_Intraday.csv',
 'BANKNIFTY_20200804_Intraday.csv',
 'BANKNIFTY_20200805_Intraday.csv',
 'BANKNIFTY_20200807_Intraday.csv',
 'BANKNIFTY_20200810_Intraday.csv',
 'BANKNIFTY_20200811_Intraday.csv',
 'BANKNIFTY_20200812_Intraday.csv',
 'BANKNIFTY_20200814_Intraday.csv',
 'BANKNIFTY_20200817_Intraday.csv',
 'BANKNIFTY_20200818_Intraday.csv',
 'BANKNIFTY_20200819_Intraday.csv',
 'BANKNIFTY_20201216_Intraday.csv',
 'BANKNIFTY_20201218_Intraday.csv',
 'BANKNIFTY_20201221_Intraday.csv',
 'BANKNIFTY_20201223_Intraday.csv']

In [55]:
save_df.loc['2020-07-30']

filename                              BANKNIFTY_20200730_Intraday.csv
Strike              [18500.0, 18600.0, 18700.0, 18800.0, 18900.0, ...
ExpiryDate                       [2020-07-30, 2020-08-06, 2020-08-13]
number_of_strike                                                  119
number_of_expiry                                                    3
Name: 2020-07-30 00:00:00, dtype: object

In [180]:
save_df.loc['2020-10-29']

filename                              BANKNIFTY_20201029_Intraday.csv
Strike              [18600.0, 18700.0, 18800.0, 18900.0, 19000.0, ...
ExpiryDate                                               [2020-10-29]
number_of_strike                                                  104
number_of_expiry                                                    1
Name: 2020-10-29 00:00:00, dtype: object

 Note : nearest_monthly does not exists '20200731'

In [53]:
save_df

filename  \
Date                                          
2020-01-01  BANKNIFTY_20200101_Intraday.csv   
2020-01-02  BANKNIFTY_20200102_Intraday.csv   
2020-01-03  BANKNIFTY_20200103_Intraday.csv   
2020-01-07  BANKNIFTY_20200107_Intraday.csv   
2020-01-08  BANKNIFTY_20200108_Intraday.csv   
...                                     ...   
2020-12-21  BANKNIFTY_20201221_Intraday.csv   
2020-12-23  BANKNIFTY_20201223_Intraday.csv   
2020-12-28  BANKNIFTY_20201228_Intraday.csv   
2020-12-29  BANKNIFTY_20201229_Intraday.csv   
2020-12-30  BANKNIFTY_20201230_Intraday.csv   

                                                       Strike  \
Date                                                            
2020-01-01  [27500.0, 27600.0, 27700.0, 27800.0, 27900.0, ...   
2020-01-02  [27500.0, 27600.0, 27700.0, 27800.0, 27900.0, ...   
2020-01-03  [28500.0, 28600.0, 28700.0, 28800.0, 28900.0, ...   
2020-01-07  [28500.0, 28600.0, 28700.0, 28800.0, 28900.0, ...   
2020-01-08  [28500.0, 28600.0, 28700.0, 28800.0, 28900.0, ...   
...                                                       ...   
2020-12-21  [21500.0, 21600.0, 21700.0, 21800.0, 21900.0, ...   
2020-12-23  [21500.0, 21600.0, 21700.0, 21800.0, 21900.0, ...   
2020-12-28  [17500.0, 17600.0, 17700.0, 17800.0, 17900.0, ...   
2020-12-29  [17500.0, 17600.0, 17700.0, 17800.0, 17900.0, ...   
2020-12-30  [17500.0, 17600.0, 17700.0, 17800.0, 17900.0, ...   

                                                   ExpiryDate  \
Date                                                            
2020-01-01  [2020-01-02, 2020-01-30, 2020-01-09, 2020-01-1...   
2020-01-02  [2020-01-02, 2020-01-30, 2020-01-09, 2020-01-1...   
2020-01-03  [2020-01-30, 2020-01-09, 2020-01-16, 2020-01-2...   
2020-01-07  [2020-01-30, 2020-01-09, 2020-01-16, 2020-01-2...   
2020-01-08  [2020-01-30, 2020-01-09, 2020-01-16, 2020-01-2...   
...                                                       ...   
2020-12-21                                       [2020-12-24]   
2020-12-23                                       [2020-12-24]   
2020-12-28                                       [2020-12-31]   
2020-12-29                                       [2020-12-31]   
2020-12-30                                       [2020-12-31]   

           number_of_strike number_of_expiry  
Date                                          
2020-01-01               80               10  
2020-01-02               80               10  
2020-01-03               78                9  
2020-01-07               78                9  
2020-01-08               78                9  
...                     ...              ...  
2020-12-21              130                1  
2020-12-23              130                1  
2020-12-28              164                1  
2020-12-29              165                1  
2020-12-30              169                1  

[206 rows x 5 columns]

In [ ]:
# calculate gamma
# exchtoken : 35788 is present at 2020-01-01 but not present in 2020-01-02 and 2020-01-03

In [141]:
holiday_2020 = market_holiday(holiday_url_2020)

### Missing Day


Total Days 365 - trading days  should match with holiday + saturday sunday 

In [149]:
all_dates = pd.date_range(start='2020-01-01', end = '2020-12-31')

In [152]:
dates

DatetimeIndex(['2020-01-01', '2020-01-02', '2020-01-03', '2020-01-06',
               '2020-01-07', '2020-01-08', '2020-01-09', '2020-01-10',
               '2020-01-13', '2020-01-14',
               ...
               '2020-12-14', '2020-12-15', '2020-12-16', '2020-12-17',
               '2020-12-18', '2020-12-21', '2020-12-23', '2020-12-28',
               '2020-12-29', '2020-12-30'],
              dtype='datetime64[ns]', name='Date', length=238, freq=None)

In [156]:
len(holiday_2020)

12

In [150]:
# def missing days():
store_date = []
dates = save_df.index
for date in all_dates:
    if date not in dates:
        store_date.append(date)   

In [154]:
len(store_date)

128

In [155]:
len(dates)

238

`NOTES` : 2020 no missing days(trading days (238) + holiday(12) + saturday_sunday(108) = 366days)

### Holiday

In [130]:
holiday_url_2020 = 'https://www.5paisa.com/stock-market-holidays-2020'
holiday_url_2021 = 'https://www.5paisa.com/stock-market-holidays-2021'
holiday_url_2022 = 'https://www.5paisa.com/stock-market-holidays-2022'
holiday_url_2023 = 'https://www.5paisa.com/stock-market-holidays-2023'
def market_holiday(url:str) -> list:
    
    market_holiday = pd.read_html(url)[0]
    market_holiday['Date'] = pd.to_datetime(market_holiday['Date'])
    market_holiday.set_index('Date', inplace = True)
    holiday_list = market_holiday.index.to_list()
    return holiday_list

In [131]:
holiday_2020 = market_holiday(holiday_url_2020)
holiday_2021 = market_holiday(holiday_url_2021)
holiday_2022 = market_holiday(holiday_url_2022)
holiday_2023 = market_holiday(holiday_url_2023)

In [133]:
len(holiday_2020)

12

### Log Time 

In [32]:
old_file_names = glob.glob('/home/cloudcraftz/Github/Log/*.csv')
old_file_names

['/home/cloudcraftz/Github/Log/log_timings.csv',
 '/home/cloudcraftz/Github/Log/log_timings_wo_debug.csv']

In [165]:
for file in old_file_names:
    print(file)
    data = pd.read_csv(file)
    columns = ['Date Time', 'extra']
    data = data.shift(1)
    first_row_values = data.columns.to_list()
    data.loc[0] = first_row_values
    data.columns = columns
    data['Date Time'] = data['Date Time'].apply(lambda x:x[2:-1])
    data.drop('extra', axis = 1, inplace = True)

    break

/home/cloudcraftz/Github/Log/log_timings.csv


In [166]:
# data['Date Time'] = data['ate Time'].astype(str)

In [167]:
# data.set_index('Date Time', inplace = True)

In [170]:
# dates = data.index
# data['Date'] = np.nan
# for idx in range(len(data)-1):
#     print(idx)
#     data.loc[idx,'Date'] = [pd.to_datetime([dates[idx]])]
# #     break

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
27

1911
1912
1913
1914
1915
1916
1917
1918
1919
1920
1921
1922
1923
1924
1925
1926
1927
1928
1929
1930
1931
1932
1933
1934
1935
1936
1937
1938
1939
1940
1941
1942
1943
1944
1945
1946
1947
1948
1949
1950
1951
1952
1953
1954
1955
1956
1957
1958
1959
1960
1961
1962
1963
1964
1965
1966
1967
1968
1969
1970
1971
1972
1973
1974
1975
1976
1977
1978
1979
1980
1981
1982
1983
1984
1985
1986
1987
1988
1989
1990
1991
1992
1993
1994
1995
1996
1997
1998
1999
2000
2001
2002
2003
2004
2005
2006
2007
2008
2009
2010
2011
2012
2013
2014
2015
2016
2017
2018
2019
2020
2021
2022
2023
2024
2025
2026
2027
2028
2029
2030
2031
2032
2033
2034
2035
2036
2037
2038
2039
2040
2041
2042
2043
2044
2045
2046
2047
2048
2049
2050
2051
2052
2053
2054
2055
2056
2057
2058
2059
2060
2061
2062
2063
2064
2065
2066
2067
2068
2069
2070
2071
2072
2073
2074
2075
2076
2077
2078
2079
2080
2081
2082
2083
2084
2085
2086
2087
2088
2089
2090
2091
2092
2093
2094
2095
2096
2097
2098
2099
2100
2101
2102
2103
2104
2105
2106
2107
2108
2109
2110


KeyboardInterrupt: 

In [146]:
# pd.Series(data.index).apply(lambda x: str(x).strip()).unique()

In [149]:
data['Date Time'] = data['Date Time'].apply(lambda x: str(x)[1:].strip())
data['extra'] = data['extra'].apply(lambda x: str(x)[:-1].strip())

In [156]:
data['Date Time'].values[0]

'023-04-24 13:48:51'

In [155]:
pd.to_datetime(data['Date Time'].values[0])

OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 23-04-24 13:48:51

In [151]:
data['Date Time'] = pd.to_datetime(data['Date Time'])

OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 23-04-24 13:48:51